CppImport/src/macro/StaticStuff.cpp

```diff
@@ -0,0 +1,358 @@
+ **
+ **
+ ** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ ** INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ ** DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ ** SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ ** SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ ** WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ ** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ **
+ **********************************************************************************************************************/
+
+#include "StaticStuff.h"
+
+namespace CppImport {
+
+void StaticStuff::orderNodes(QVector<Model::Node*>& input)
+{
+	qSort(input.begin(), input.end(),
+			[](Model::Node* e1, Model::Node* e2)
+	{
+		if (auto commonAncestor = e1->lowestCommonAncestor(e2))
+			if (auto list = DCast<Model::List>(commonAncestor))
+			{
+				int index1 = -1;
+				for (auto c : list->children())
+					if (c == e1 || c->isAncestorOf(e1))
+					{
+						index1 = list->indexOf(c);
+						break;
+					}
+
+				int index2 = -1;
+				for (auto c : list->children())
+					if (c == e2 || c->isAncestorOf(e2))
+					{
+						index2 = list->indexOf(c);
+						break;
+					}
+
+				return index1 < index2;
+			}
+
+		return true;
+	});
+}
+
+bool StaticStuff::validContext(Model::Node* node)
+{
+	if (DCast<OOModel::Project>(node))
+		return true;
+	else if (DCast<OOModel::Module>(node))
+		return true;
+	else if (DCast<OOModel::Class>(node))
+		return true;
+	else if (DCast<OOModel::Method>(node))
+		return true;
+	else
+		return false;
+}
+
+OOModel::Declaration*StaticStuff::actualContext(Model::Node* node)
+{
+	auto current = node->parent();
+
+	while (current)
+	{
+		if (auto result = DCast<OOModel::Project>(current))
+			return result;
+		else if (auto result = DCast<OOModel::Module>(current))
+			return result;
+		else if (auto result = DCast<OOModel::Class>(current))
+			return result;
+		else if (auto result = DCast<OOModel::Method>(current))
+			return result;
+		else
+			current = current->parent();
+	}
+
+	Q_ASSERT(false);
+}
+
+OOModel::Declaration*StaticStuff::createContext(OOModel::Declaration* actualContext)
+{
+	if (DCast<OOModel::Project>(actualContext))
+		return new OOModel::Project("Context");
+	else if (DCast<OOModel::Module>(actualContext))
+		return new OOModel::Module("Context");
+	else if (DCast<OOModel::Class>(actualContext))
+		return new OOModel::Class("Context");
+	else if (DCast<OOModel::Method>(actualContext))
+		return new OOModel::Method("Context");
+
+	Q_ASSERT(false);
+}
```

Comment:
For this method and some of the previous ones, I imagine that there is one place where they are actually used. Is that so? If true, then why not put these methods directly there?